lib/sequent/core/event_publisher.rb

```diff
@@ -0,0 +1,57 @@
+module Sequent
+  module Core
+    #
+    # EventPublisher ensures that, for every thread, events will be published in the order in which they are meant to be published.
+    #
+    # This potentially introduces a wrinkle into your plans: You therefore should not split a "unit of work" across multiple threads.
+    #
+    # If you do not want this, you are free to implement your own version of EventPublisher and configure sequent to use it.
+    #
+    class EventPublisher
+      class PublishEventError < RuntimeError
+        attr_reader :event_handler_class, :event
+
+        def initialize(event_handler_class, event)
+          @event_handler_class = event_handler_class
+          @event = event
+        end
+
+        def message
+          "Event Handler: #{@event_handler_class.inspect}\nEvent: #{@event.inspect}\nCause: #{cause.inspect}"
+        end
+      end
+
+      def publish_events(events)
+        return if configuration.disable_event_handlers
+        events.each { |event| events_queue.push(event) }
+        process_events
+      end
+
+      private
+
+      def events_queue
+        Thread.current[:events_queue] ||= Queue.new
```

Comment:
Queue only supports FIFO I think. We can safely use Array here to support both FIFO and LIFO.

Reply:
I can't imagine why we would want to support LIFO.

Reply:
We are changing the current behavior so I think it is good to support both cases and the current one as default, with maybe a warning message.

Reply:
Neither of those scenarios is the current one.

Reply:
Yes you are right... Okay, let's communicate this in the changelog when we release.